e2e/cli/command/provision.go

```diff
@@ -0,0 +1,275 @@
+	cmdFlags.Usage = func() { log.Println(c.Help()) }
+	cmdFlags.StringVar(&envPath, "env-path", "./environments/", "Path to e2e environment terraform configs")
+	cmdFlags.StringVar(&nomadBinary, "nomad-binary", "", "")
+	cmdFlags.BoolVar(&destroy, "destroy", false, "")
+	cmdFlags.StringVar(&tfPath, "tf-path", "", "")
+
+	if err := cmdFlags.Parse(args); err != nil {
+		log.Fatalf("failed to parse flags: %v", err)
+	}
+
+	args = cmdFlags.Args()
+	if len(args) != 2 {
+		log.Fatalf("expected 2 args, but got: %v", args)
+		log.Fatal(c.Help())
+	}
+
+	env, err := newEnv(envPath, args[0], args[1], tfPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if destroy {
+		if err := env.destroy(); err != nil {
+			log.Fatal(err)
+			return 1
+		}
+		fmt.Println("Environment successfully destroyed")
+		return 0
+	}
+
+	// Use go-getter to fetch the nomad binary
+	nomadPath, err := c.fetchBinary(nomadBinary)
+	defer os.RemoveAll(nomadPath)
+
+	results, err := env.provision(nomadPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fmt.Printf(strings.TrimSpace(`
+NOMAD_ADDR=%s
+	`), results.nomadAddr)
+
+	return 0
+}
+
+// Fetches the nomad binary and returns the temporary directory where it exists
+func (c *Provision) fetchBinary(bin string) (string, error) {
+	nomadBinaryDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return "", fmt.Errorf("failed to create temp dir: %v", err)
+	}
+
+	if err = getter.GetFile(path.Join(nomadBinaryDir, "nomad"), bin); err != nil {
+		return "", fmt.Errorf("failed to get nomad binary: %v", err)
+	}
+
+	return nomadBinaryDir, nil
+}
+
+type environment struct {
+	path     string
+	provider string
+	name     string
+
+	tf      string
+	tfPath  string
+	tfState string
+}
+
+type envResults struct {
+	nomadAddr string
+}
+
+func newEnv(envPath, provider, name, tfStatePath string) (*environment, error) {
+	// Make sure terraform is on the PATH
+	tf, err := exec.LookPath("terraform")
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup terraform binary: %v", err)
+	}
+
+	// set the path to the terraform module
+	tfPath := path.Join(envPath, provider, name)
+	log.Printf("[DEBUG] provision: using tf path %s", tfPath)
+	if _, err := os.Stat(tfPath); os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to lookup terraform configuration dir %s: %v", tfPath, err)
+	}
+
+	// set the path to state file
+	tfState := path.Join(tfStatePath, fmt.Sprintf("e2e.%s.%s.tfstate", provider, name))
+
+	env := &environment{
+		path:     envPath,
+		provider: provider,
+		name:     name,
+		tf:       tf,
+		tfPath:   tfPath,
+		tfState:  tfState,
+	}
+	return env, nil
+}
+
+// provision calls terraform to setup the environment with the given nomad binary
+func (env *environment) provision(nomadPath string) (*envResults, error) {
+	tfArgs := []string{"apply", "-auto-approve", "-input=false", "-no-color",
+		"-state", env.tfState,
+		"-var", fmt.Sprintf("nomad_binary=%s", path.Join(nomadPath, "nomad")),
```

Comment:
We also need to pass the SSH key name for ec2 to terraform. Or is the expectation that people modify their local checkout of `./environment` to include that in their tfvars file?

Reply:
Chatted about this out of band. Plan is to use the [`tls_private_key`](https://www.terraform.io/docs/providers/tls/r/private_key.html) terraform resource to generate a keypair dynamically which could be fetched from the tfstate if needed.